Custom subway icons

starsep opened the pull request:
See #3509

biodranik reviewed (commented):
@vng can't we map icons directly to the value of the network tag? Then we don't need any additional code for it. Our icons system should support even Cyrillic names for icons, no?
```
    /// @todo Probably, we can delete this processing because cities
    /// are matched by limit rect in MatchCity.
    if (!subwayDone && types.IsSubwayStation(vTypes[i]))
    {
      TagProcessor(p).ApplyRules({
          {"network", "London Underground", [&params] { params.SetRwSubwayType("london"); }},
          {"network", "New York City Subway", [&params] { params.SetRwSubwayType("newyork"); }},
          {"network", "Московский метрополитен", [&params] { params.SetRwSubwayType("moscow"); }},
          {"network", "Петербургский метрополитен", [&params] { params.SetRwSubwayType("spb"); }},
          {"network", "Verkehrsverbund Berlin-Brandenburg",
           [&params] { params.SetRwSubwayType("berlin"); }},
          {"network", "Минский метрополитен", [&params] { params.SetRwSubwayType("minsk"); }},

          {"network", "Київський метрополітен", [&params] { params.SetRwSubwayType("kiev"); }},
          {"operator", "КП «Київський метрополітен»",
           [&params] { params.SetRwSubwayType("kiev"); }},

          {"network", "RATP", [&params] { params.SetRwSubwayType("paris"); }},
          {"network", "Metro de Barcelona", [&params] { params.SetRwSubwayType("barcelona"); }},

          {"network", "Metro de Madrid", [&params] { params.SetRwSubwayType("madrid"); }},
          {"operator", "Metro de Madrid", [&params] { params.SetRwSubwayType("madrid"); }},

          {"network", "Metropolitana di Roma", [&params] { params.SetRwSubwayType("roma"); }},
          {"network", "ATAC", [&params] { params.SetRwSubwayType("roma"); }},
      });
```

starsep commented:
Another thing to tackle which might be fixed by changing approach are redundant translations: https://github.com/organicmaps/organicmaps/blob/master/data/strings/types_strings.txt#L17132-L17504

biodranik commented:
Translations now can use ref = to the base type, to avoid copy-pasting.

vng commented:
@biodranik Well, MatchCity by rect is more universal way. That's why I put todo to remove matching by network code.
network values are always _fancy_ and should be added manually for each city.

@starsep 
 1. Now I see 2 Warsaw icons in repo: subway-warsaw-*.svg and subway-warszawa-*.svg AFAIR, _warszawa_ is the correct one, but what is the _warsaw_ icon about? Is it useful or we should delete it?
 2. I usually recommend to use some existing _deprecated_ entries when editing mapcss-mapping.csv instead of adding the new ones .. Near with existing subway entries.
 3. You should manually edit 2 files here: mapcss-mapping.csv and Subways.mapcss. Other files are generated manually by generate_styles.sh tool See https://github.com/organicmaps/organicmaps/blob/master/docs/STYLES.md for details.
And also you should create 2 separate commits: 1 - Your edits. 2 - Styles regenerated commit.

starsep commented:
1. warsaw icon is of WKD https://en.wikipedia.org/wiki/Warsaw_Commuter_Railway, light rail commuter line.
I will handle 2. and 3. in the evening

vng commented:
Hm, Subways.mapcss already have needed rules, we have needed icons.
The only thing is to add new types into mapcss-mapping.csv
Should we add only Warsaw, or all available cities with a bulk change ;)

starsep commented:
I made Warsaw change in two commits. I regenerated mwm file and icons work correctly.

In Subways.mapcss there are 105 cities.
In mapcss-mapping.csv there 12 cities.

Bulk edit would mean extra 2*(105-12)=186 types
At the moment there are 374 `deprecated|deprecated` types, so it's enough to only replace deprecated one.
I can work on a bulk change.
Is it important to keep subway types close together? I can take first 186 first deprecated types for example.

Just speculating, please correct me if I am wrong: number in mapcss-mapping.csv is used as some kind of type id in mwm file. And it cannot be changed to keep backwards compatibilty with old map files.

Rebuilding index only i.e. `generator_tool --generate_index=true` wouldn't work here to test all cities affected, right?
To test every city I should either regenerate all map files (takes a ton of time) or make a final testing just before release once new maps are generated?

biodranik commented:
I'll throw my 5 cents here: how many unique transport lines (not only the metro!) with unique icons are there in the world? Is it the best approach to manually hardcode them in types? Isn't it prone to mistakes?

What about other icons unique per country level (e.g. a driving training school sign)? What approach can be applied there? What other per-country icons are there?

Are there any alternative approaches? E.g. store the type/icon/network in the feature's metadata?

vng commented:
@biodranik 
- At this moment we have only classifier type simple mechanics to draw icons. I don't see problems to make many types here.
- Can't say is good or not to have unique-per-country icons like driving schools .. If yes, we should invent some other mechanics and it's a long story ..
- Metadata is not as good as classifier type for fast fetching, but it's used now in 3D-buildings.

starsep commented:
I pushed commit with missing city subway types (and styles regenerated afterwards). Generated with:
```python
MISSING = [
    "adana",
    "algiers",
    "almaty",
    "amsterdam",
    "ankara",
    "athens",
    "baku",
    "bangkok",
    "beijing",
    "bengalore",
    "bilbao",
    "brasilia",
    "brescia",
    "brussels",
    "bucharest",
    "budapest",
    "buenos_aires",
    "bursa",
    "cairo",
    "caracas",
    "catania",
    "changchun",
    "chengdu",
    "chicago",
    "chongqing",
    "dalian",
    "delhi",
    "dnepro",
    "dubai",
    "ekb",
    "fukuoka",
    "glasgow",
    "guangzhou",
    "hamburg",
    "helsinki",
    "hiroshima",
    "isfahan",
    "istanbul",
    "izmir",
    "kazan",
    "kharkiv",
    "kobe",
    "kolkata",
    "kunming",
    "kyoto",
    "la",
    "lausanne",
    "lille",
    "lima",
    "lisboa",
    "lisbon",
    "lyon",
    "malaga",
    "manila",
    "maracaibo",
    "mashhad",
    "mecca",
    "medellin",
    "mexico",
    "milan",
    "montreal",
    "munchen",
    "nagoya",
    "nnov",
    "novosibirsk",
    "osaka",
    "oslo",
    "palma",
    "panama",
    "philadelphia",
    "pyongyang",
    "rennes",
    "rio",
    "rotterdam",
    "samara",
    "santiago",
    "santo_domingo",
    "saopaulo",
    "sapporo",
    "sendai",
    "sf",
    "shanghai",
    "shiraz",
    "sofia",
    "stockholm",
    "tabriz",
    "taipei",
    "taoyuan",
    "tashkent",
    "tbilisi",
    "tianjin",
    "tokyo",
    "valencia",
    "vienna",
    "washington",
    "wuhan",
    "yerevan",
    "yokohama",
]

with open("data/mapcss-mapping.csv", "r") as f:
    mappings = f.readlines()

station = True
cityId = 0
with open("data/mapcss-mapping.csv", "w") as f:
    for mapping in mappings:
        if "deprecated" not in mapping or (cityId == len(MISSING) and not station):
            f.write(mapping)
            continue
        if cityId == len(MISSING):
            cityId = 0
            station = False
        typeId = int(mapping.split(";")[1])
        city = MISSING[cityId]
        cityId += 1
        if station:
            f.write(
                f"railway|station|subway|{city};[railway=station][transport=subway][city={city}],[railway=station][station=subway][city={city}];;name;int_name;{typeId};\n"
            )
        else:
            f.write(
                f"railway|subway_entrance|{city};[railway=subway_entrance][city={city}];;name;int_name;{typeId};\n"
            )
```

vng commented:
Holy moly ;) Your efforts are impressive :)

I'd like to keep deprecates numbers from 0 to 700 reserved for more _useful_ types. Generic rule is less number -> more frequent type. Can you put this subways starting from say deprecated id = 747 and below ..

@biodranik @rtsisyk Are you ok if we will make bulk insert of all existing city metro icon types?

vng commented:
Or even put this types to the bottom of mapcss-mapping.csv. I couldn't imagine that you will make this bulk insert ;)

starsep commented:
Why lower ids are reserved for useful types? Just a convention or is some dynamic length binary coding used?

I put them on the end, generated with:
```python
MISSING = [
    "adana",
    "algiers",
    "almaty",
    "amsterdam",
    "ankara",
    "athens",
    "baku",
    "bangkok",
    "beijing",
    "bengalore",
    "bilbao",
    "brasilia",
    "brescia",
    "brussels",
    "bucharest",
    "budapest",
    "buenos_aires",
    "bursa",
    "cairo",
    "caracas",
    "catania",
    "changchun",
    "chengdu",
    "chicago",
    "chongqing",
    "dalian",
    "delhi",
    "dnepro",
    "dubai",
    "ekb",
    "fukuoka",
    "glasgow",
    "guangzhou",
    "hamburg",
    "helsinki",
    "hiroshima",
    "isfahan",
    "istanbul",
    "izmir",
    "kazan",
    "kharkiv",
    "kobe",
    "kolkata",
    "kunming",
    "kyoto",
    "la",
    "lausanne",
    "lille",
    "lima",
    "lisboa",
    "lisbon",
    "lyon",
    "malaga",
    "manila",
    "maracaibo",
    "mashhad",
    "mecca",
    "medellin",
    "mexico",
    "milan",
    "montreal",
    "munchen",
    "nagoya",
    "nnov",
    "novosibirsk",
    "osaka",
    "oslo",
    "palma",
    "panama",
    "philadelphia",
    "pyongyang",
    "rennes",
    "rio",
    "rotterdam",
    "samara",
    "santiago",
    "santo_domingo",
    "saopaulo",
    "sapporo",
    "sendai",
    "sf",
    "shanghai",
    "shiraz",
    "sofia",
    "stockholm",
    "tabriz",
    "taipei",
    "taoyuan",
    "tashkent",
    "tbilisi",
    "tianjin",
    "tokyo",
    "valencia",
    "vienna",
    "washington",
    "wuhan",
    "yerevan",
    "yokohama",
]

station = True
cityId = 0
typeId = 1356
with open("data/mapcss-mapping.csv", "a") as f:
    for station in [True, False]:
        for city in MISSING:
            if station:
                f.write(
                    f"railway|station|subway|{city};[railway=station][transport=subway][city={city}],[railway=station][station=subway][city={city}];;name;int_name;{typeId};\n"
                )
            else:
                f.write(
                    f"railway|subway_entrance|{city};[railway=subway_entrance][city={city}];;name;int_name;{typeId};\n"
                )
            typeId += 1
```

starsep commented:
Test failed: https://github.com/organicmaps/organicmaps/actions/runs/3185151162/jobs/5194386832
> 2022-10-04 20:37:10,822 The server's life has come to an end, pid: 10682

At first I thought tests take longer now due to extra types but previous jobs take even more time. I guess it's unrelated to my change.

vng commented:
Failed test say that:
- no _clear_ icons for [7: subway-bilbao-m subway-bilbao-s subway-hamburg-m subway-hamburg-s subway-lausanne-s subway-munchen-m subway-munchen-s ]
- no _night_ icons for [20: subway-bilbao-m subway-bilbao-s subway-hamburg-m subway-hamburg-s subway-lausanne-m subway-lausanne-s subway-lille-m subway-lille-s subway-lima-m subway-lima-s subway-lisboa-m subway-lisboa-s subway-losangeles-m subway-losangeles-s subway-lyon-m subway-lyon-s subway-munchen-m subway-munchen-s subway-vienna-m subway-vienna-s ]

- I suppose that night icons was made from clear ones using some color/opacity transformation.
- Is it possible to get Hamburg or Munchen svg metro icons somewhere? Why did you include this cities into your list?

starsep commented:
I simply took all missing cities with styling defined in https://github.com/organicmaps/organicmaps/blob/master/data/styles/clear/include/Subways.mapcss
I will try to find those icons, if I won't be able to do so I will remove those types.

AlertSubject commented:
> There is general icon U-Bahn in Germany
> https://commons.m.wikimedia.org/wiki/File:U-Bahn.svg
> @AlertSubject perhaps you can find more specific icons or is this general one used in those cities? I think it is in Hamburg, I haven't been to Munich.
> 

Yeah this is indeed the one for Hamburg
But I've never seen it used on the map
![Screenshot_2022-10-05-09-29-48-63_75d0290a7c00525acd83c0ab8700d64c.jpg](https://user-images.githubusercontent.com/72819077/194004875-2b419648-938b-40a8-b399-00ab0b3479f5.jpg)

This is a U-Bahn Station in Hamburg and there is no specific icon displayed sadly
Neither is the S-Bahn icon
And I believe the S-Bahn icon is used all over Germany as they are operated by Deutsche Bahn

vng commented:
So make sense to rename subway-berlin icon to subway-ubahn and add Berlin, Munchen, Hamburg types manually.

AlertSubject commented:
https://commons.m.wikimedia.org/wiki/Category:S-Bahn_logos

Here's a list of some German S-Bahn icons
Hamburg is in there as well for example
Berlin too

starsep commented:
I renamed subway-wien to subway-vienna. I made night version icons by transforming clear (day) ones.
I made ubahn icon out wikimedia svg and used it for Ham

To get night icon from clear icon: opacity is changed from 0.6 to 0.4 and color is multiplied by some factor ~0.6
```python
def night(rgb: str) -> str:
    r, g, b = int(rgb[:2], 16), int(rgb[2:4], 16), int(rgb[4:6], 16)
    transform = lambda x: (hex(int(x * 0.6))[2:]).zfill(2).upper()
    return transform(r) + transform(g) + transform(b)
# night("1565C0") == "0C3C73"
```

missed [3: subway-bilbao-m subway-bilbao-s subway-lausanne-s ]

starsep commented:
I don't know what to do with Metro Bilbao Logo: https://en.wikipedia.org/wiki/File:Logo_Metro_Bilbao.svg
I will make subway-lausanne-s by transforming subway-lausanne-m

vng commented:
- Aren't subway-berlin and new subway-ubahn the same icons?
- Why subway-ubahn-m and subway-ubahn-s have different background colors?

vng commented:
Also subway-lausanne-m.svg doesn't have corresponding -s version ..

vng commented:
I think, we can just rename berlin icong to ubahn, no?

starsep commented:
> * Aren't subway-berlin and new subway-ubahn the same icons?
Almost the same, I can remove subway-berlin and use subway-ubahn for Berlin.

>     * Why subway-ubahn-m and subway-ubahn-s have different background colors?
They seem to have the same background color

> I think, we can just rename berlin icon to ubahn, no?
Sure, I can use Berlin icon as ubahn as well

starsep commented:
> > Why subway-ubahn-m and subway-ubahn-s have different background colors?
> 
> They seem to have the same background color

I missed opacity, I will fix that

vng commented:
To clarify, github shows me this:
<img width="421" alt="Screenshot 2022-10-05 at 22 41 49" src="https://user-images.githubusercontent.com/175612/194148357-384d3a15-1bc2-432b-ab80-eee8d9e41c0f.png">

I think, better to rename existing Berlin, instead of adding the new one.

starsep commented:
Berlin icon renamed to U-Bahn, subway-lausanne-s created from subway-lausanne-m.
Only Bilbao missing: https://en.wikipedia.org/wiki/File:Logo_Metro_Bilbao.svg

vng commented on data/styles/clear/include/Subways.mapcss:
```diff
@@ -587,18 +587,18 @@ node|z17[railway=subway_entrance][city=montreal] {icon-image: subway-montreal-s.
 node|z14[railway=station][transport=subway][city=nagoya] {icon-image: subway-nagoya-s.svg;}
 node|z15-16[railway=station][transport=subway][city=nagoya] {icon-image: subway-nagoya-m.svg;}
-node|z17-[railway=station][transport=subway][city=nagoya] {icon-image: subway-nagoya-m.svg;text-optional: true;}
+node|z17-[railway=station][transport=subway][city=nagoya] {ion-image: subway-nagoya-m.svg;text-optional: true;}
```
Bad change here. "ion"

starsep replied:
Thanks for spotting that, fixed!

vng commented:
Great! Let see what CI checks will show. Also:
1. could you please make script change for new subway strings (types_strings.txt) like:
```
  [type.railway.station.subway.CITY]
    ref = type.railway.station.subway
```
2. Rebase you branch on fresh master (where your previous merged subway strings) and squash commits on 3 like: 
- Added subway icons for cities
- [styles] Regenerated
- [strings] Regenerated

biodranik commented:
Can you please show several screenshots of how it looks now on the map? I'm afraid the map becomes less readable/intuitive. Of course, all local people know metro symbols. But what about travellers? When you come into some city, and want to find the closest metro on the map, how can you understand that it's some, say, 'S' symbol?

starsep commented:
I can make some screenshots tomorrow. I think most of them are variants of M. I didn't add S Bahn icon.
As a traveller I would first turn on subway layer to see what lines are in the city. I think those icons allow to find subway station nearby easier without turning a layer or searching.

matheusgomesms commented:
For Bilbao, you can use just the red rings. I've lived there and it is [the only logo on metro entrances.](https://www.google.com/maps/@43.2632459,-2.9343872,3a,75y,154.25h,94.73t/data=!3m6!1e1!3m4!1sI-EcZ9w2DPIYzgWZt5BQTA!2e0!7i16384!8i8192)

AlertSubject commented:
> Can you please show several screenshots of how it looks now on the map? I'm afraid the map becomes less readable/intuitive. Of course, all local people know metro symbols. But what about travellers? When you come into some city, and want to find the closest metro on the map, how can you understand that it's some, say, 'S' symbol?


The S symbol on green background is found in all the train stations and all the trains. It's shown on all the boards that navigate you to the train platforms
I'm pretty sure a visitor will quickly recognize the familiar symbol and be glad to see it on the map

matheusgomesms reviewed (approved):
I've reviewed and fixed Portuguese strings. Looks good to me!

vng commented:
@biodranik You can check it now for some major cities:
![photo_2022-10-06 08 47 29](https://user-images.githubusercontent.com/175612/194224239-78c1f170-c900-4362-8fb6-9e9f5240409d.jpeg)
![photo_2022-10-06 08 47 38](https://user-images.githubusercontent.com/175612/194224243-81fbd320-db33-445d-b72f-4c563cd53b0c.jpeg)
![photo_2022-10-06 08 47 42](https://user-images.githubusercontent.com/175612/194224247-9dc85750-6c4e-4d33-9551-b3fbdda7e3a6.jpeg)
![photo_2022-10-06 08 47 45](https://user-images.githubusercontent.com/175612/194224249-1df41402-e09e-4174-9bc2-e94c2a4307d1.jpeg)

Moreover, I'd prefer to draw small metro icons 1 zoom level before. Look how good it looks in Moscow, want to tap this icon first ;)
![photo_2022-10-06 08 47 35](https://user-images.githubusercontent.com/175612/194224494-7562d2e4-e58b-48c0-90ec-b0c129db7541.jpeg)

vng commented:
Moreover, I'd prefer to draw small metro icons 1 zoom level earlier. Look how good it looks in Moscow, want to tap this icon first ;)
![photo_2022-10-06 08 47 35](https://user-images.githubusercontent.com/175612/194224494-7562d2e4-e58b-48c0-90ec-b0c129db7541.jpeg)

starsep commented:
@matheusgomesms
> I've reviewed and fixed Portuguese strings. Looks good to me!

You should modify only `type.railway.station.subway` and `type.railway.subway_entrance` in a file data/strings/types_strings.txt
All `type.railway.station.subway.$city` and `type.railway.subway_entrance.$city` strings use the value of base type. All strings.txt (for Android) and Localizable.strings (for iOS) are generated automatically. See https://github.com/organicmaps/organicmaps/blob/master/docs/TRANSLATIONS.md for more details. 
I am sorry that you lost some time, I will apply your suggestions in the evening. I had no intention of asking translations-* groups for review, they have been added automatically due to modification of relevant files.

@vng thanks for screenshots!

starsep commented:
Rebased and regenerated
@matheusgomesms I added your Portuguese fixes here: b4eb2df3bc7a9bbdf5a677fe16a9958749454035

matheusgomesms commented:
> Rebased and regenerated @matheusgomesms I added your Portuguese fixes here: [b4eb2df](https://github.com/organicmaps/organicmaps/commit/b4eb2df3bc7a9bbdf5a677fe16a9958749454035)

Hi @starsep thanks for that! Just wanted to point out that the PT-BR fix is missing:

`data/strings/types_strings.txt`
line 16674: pt-BR = Estação de metrô
line 17075: pt-BR = Entrada do metrô

Could you please fix that? I couldn't do it by myself, unfortunately.

LaoshuBaby commented on android/res/values-zh/strings.xml:
```diff
@@ -1361,29 +1361,227 @@
+	<string name="type.railway.station.subway.wuhan">地铁车站</string>
+	<string name="type.railway.station.subway.yerevan">地铁车站</string>
+	<string name="type.railway.station.subway.yokohama">地铁车站</string>
 	<string name="type.railway.subway_entrance">地铁入口</string>
```
In Simplified Chinese, I'd prefer use "地铁出入口" for "Subway Entrance", if we don't clarify enter or exit behavior.

LaoshuBaby reviewed (changes requested):
+ Zh-Hant: Waiting native speaker's advice
+ zh-Hans: If we don't use 2 type to distinguish enter or exit subway system, I prefer "地铁出入口"

biodranik commented:
@vng thanks for the screenshots. As I expected, the map looks a bit polluted. The good news is that it can be workarounded now by reducing the size of all metro icons on observation zoom levels.

Google draws small icons on observation zoom levels and increases their size when you zoom in, Apple draws dots until you zoom in to almost the max zoom level. Check it yourself and compare.

Metro icons on a "general", universal map style should not pollute it, we'll implement a public transport style for that later.

vng commented:
@biodranik This is a current prod screenshots and generic metro icon vs individual metro icons - "pollution" will be the same or even worse with generic icon.
@starsep Let remove Bilbao because we don't have icons for it. Also, please, take into account Chinese comments.

s8321414 commented on android/res/values-zh-rTW/strings.xml:
```diff
@@ -1277,29 +1277,227 @@
 	<string name="type.railway.station.light_rail">輕軌車站</string>
 	<string name="type.railway.station.monorail">單軌車站</string>
 	<string name="type.railway.station.subway">地鐵車站</string>
```
> @s8321414 ↑ native speaker of Traditional Chinese I guess he'll prefer "捷運車站" more than "地鐵車站", so I call him for help and waiting for advice.

For subway station, we use 「捷運站」or「捷運車站」 in Taiwan.

s8321414 commented:
> * Zh-Hant: Waiting native speaker's advice
> 
>     * zh-Hans: If we don't use 2 type to distinguish enter or exit subway system, I prefer "地铁出入口"

Just replied :-)
https://github.com/organicmaps/organicmaps/pull/3514#discussion_r989567200

matheusgomesms commented:
> @starsep Let remove Bilbao because we don't have icons for it. Also, please, take into account Chinese comments.

Please see my [comment ](https://github.com/organicmaps/organicmaps/pull/3514#issuecomment-1269007134) about Bilbao.

starsep commented:
@matheusgomesms 
I removed Bilbao for now. We could theoretically use this icon: https://commons.wikimedia.org/wiki/File:S%C3%ADmbolo_del_Metro_de_Bilbao.svg but it's really cryptic
On purpose done in separate commit: 79436fb9b5c75ce4f2a9d980a33cb9721f691c72
I added pt-BR translations in 934a04ebb8ec3075a67e8e050dfa1efa992271b8

@vng @s8321414 @LaoshuBaby 
I changed Chinese translations in 934a04ebb8ec3075a67e8e050dfa1efa992271b8

Style/symbols/strings regenerated.

vng commented:
Good! mapcss-mapping.csv should **not** renumber existing types, so separate Bilbao commit is weird. Better to squash it into first main commit.

AlertSubject commented:
> https://commons.m.wikimedia.org/wiki/Category:S-Bahn_logos
> 
> Here's a list of some German S-Bahn icons
> Hamburg is in there as well for example
> Berlin too 

Has someone added Hamburg as well?
I believe Berlin is included in this latest update
Have any other cities been added?

starsep commented:
Berlin and Hamburg use S-Bahn icon. 
Many more cities have been added: adana, algiers, almaty, amsterdam, ankara, athens, baku, bangkok, beijing, bengalore, brasilia, brescia, brussels, bucharest, budapest, buenos_aires, bursa, cairo, caracas, catania, changchun, chengdu, chicago, chongqing, dalian, delhi, dnepro, dubai, ekb, fukuoka, glasgow, guangzhou, hamburg, helsinki, hiroshima, isfahan, istanbul, izmir, kazan, kharkiv, kobe, kolkata, kunming, kyoto, la, lausanne, lille, lima, lisboa, lisbon, lyon, malaga, manila, maracaibo, mashhad, mecca, medellin, mexico, milan, montreal, munchen, nagoya, nnov, novosibirsk, osaka, oslo, palma, panama, philadelphia, pyongyang, rennes, rio, rotterdam, samara, santiago, santo_domingo, saopaulo, sapporo, sendai, sf, shanghai, shiraz, sofia, stockholm, tabriz, taipei, taoyuan, tashkent, tbilisi, tianjin, tokyo, valencia, vienna, washington, wuhan, yerevan, yokohama,

AlertSubject commented:
> Hamburg

I see the U-Bahn logo on a station that only has ubahn lines going through
But I don't see any S-Bahn icon as of now
Is this due to multiple different lines using the same station or what conditions have to be met at the osm data level?

starsep commented:
At the moment we don't have icon rending for S-Bahn.
For U-Bahn in Hamburg it's `[railway=station][transport=subway][city=hamburg]`. 
In OSM tags I think it's `railway=station` + `station=subway` + bbox for Hamburg

AlertSubject commented:
> icon rendering for S-Bahn.

Should I create a new issue and link to the logo within it?
